yast2_rdp.pm

```diff
@@ -0,0 +1,194 @@
+#          successfully set.
+# Maintainer: Jun Wang <jgwang@suse.com>
+
+use base "y2_module_guitest";
+use strict;
+use warnings;
+use testapi;
+use utils;
+use version_utils 'is_sle';
+
+my $repos = {'12-SP1' => 'http://download.suse.de/ibs/Devel:/YaST:/SLE-12-SP1/SLE-12-SP1/',
+    '12-SP1-UP' => 'http://download.suse.de/ibs/Devel:/YaST:/SLE-12-SP1/SLE-12-SP1_Update/',
+    '12-SP2'    => 'http://download.suse.de/ibs/Devel:/YaST:/SLE-12-SP2/SLE_12_SP2/',
+    '12-SP2-UP' => 'http://download.suse.de/ibs/Devel:/YaST:/SLE-12-SP2/SUSE_SLE-12-SP2_Update/',
+    '12-SP3'    => 'http://download.suse.de/ibs/Devel:/YaST:/SLE-12-SP3/SLE_12_SP3/',
+    '12-SP3-UP' => '',
+    '12-SP4'    => 'http://download.suse.de/ibs/Devel:/YaST:/SLE-12-SP4/SLE_12_SP4/',
+    '12-SP4-UP' => 'http://download.suse.de/ibs/Devel:/YaST:/SLE-12-SP4/SUSE_SLE-12-SP4_Update/'
+};
+
+sub becomeroot {
+    type_string "su -\n";
+    wait_still_screen;
+    type_string "$testapi::password";
+    wait_screen_change { send_key "ret" };
+
+    script_run('test $(id -u) -eq 0 && echo "imroot" > /dev/' . $testapi::serialdev, 0);
+    wait_serial("imroot", 5) || die "Root prompt not there";
+}
+
+sub setup_system {
+    x11_start_program('xterm', valid => 0);
+
+    becomeroot;
+    # make sure that yast2-rdp has been installed.
+    if (is_sle('>12-SP2') and is_sle('<15')) {
+        my $version = get_var('VERSION');
+        zypper_ar("$repos->{$version}", 'Devel-YaST-SLE');
+        if ("$version" ne '12-SP3') {
+            zypper_ar("$repos->{qq($version-UP)}", 'Devel-YaST-SLE-UP');
+        }
+    } elsif (is_sle('<12-SP3')) {
+        die "This version of product is NOT supported.";
+    }
+    zypper_call("in yast2-rdp", exitcode => [0, 102, 103]);
+    type_string "exit\n";
+
+    # disable screen saver.
+    if (check_var("DESKTOP", "gnome")) {
+        script_run 'gsettings set org.gnome.desktop.session idle-delay 0';
+    } else {
+        script_run("xscreensaver-command -exit");
+        if (check_var("DESKTOP", "lxde")) {
+            # Disable xscreensaver autostart on LXDE (from system wide config and user config)
+            script_sudo('sed -i "s/\@xscreensaver -no-splash//" /etc/xdg/lxsession/LXDE/autostart');
+            script_run('sed -i "s/\@xscreensaver -no-splash//" ~/.config/lxsession/LXDE/autostart');
+        }
+    }
+    wait_screen_change { send_key "ctrl-d" };
+    wait_still_screen;
+}
+
+sub check_service {
+    my ($type, $cmd) = @_;
+
+    x11_start_program('xterm', valid => 0);
+    becomeroot;
+    my $result = script_output("$cmd", proceed_on_failure => 1);
+    type_string "exit\n";
+    wait_screen_change { send_key "ctrl-d" };
+    wait_still_screen 2;
+
+    if ($type eq 'General') {
+        if ($result eq "active" or $result eq "enabled") {
+            return 1;
+        } else {
+            return 0;
+        }
+    } elsif ($type eq 'Firewall') {
+        if ($result eq 'ms-wbtX') {
```

Comment:
what is `ms-wbtX`?

Reply:
ms-wbt is a return value, it means that the port of xrdp is open in firewall. X, it is just a habit for fear that "$result" is null